game_patch/multi/server.cpp

```diff
@@ -319,27 +319,37 @@ static void send_private_message_with_stats(rf::Player* player)
 void shuffle_level_array()
 {
     std::ranges::shuffle(rf::netgame.levels, rng);
-    xlog::info("Shuffled level rotation");
+    xlog::warn("Shuffled level rotation");
 }
 
 const char* get_rand_level_filename()
 {
     std::size_t num_levels = rf::netgame.levels.size();
 
+    // current_level_index keeps our position in rotation while we go somewhere else
+    const char* current_level_filename = rf::level_filename_to_load;
+
+    xlog::debug("Starting random selection process, current level is {}", current_level_filename);
+
     if (num_levels <= 1) {
         // nowhere else to go, we're staying here!
-        return rf::netgame.levels[rf::netgame.current_level_index].c_str();
+        xlog::debug("Map list has only 1 member, reloading current level {}", current_level_filename);
+        return current_level_filename;
     }
 
-    std::uniform_int_distribution<std::size_t> dist(0, num_levels - 2);
-    std::size_t rand_level_index = dist(rng);
+    std::uniform_int_distribution<std::size_t> dist_levels(0, num_levels - 1);
+    std::size_t rand_level_index = dist_levels(rng);
 
-    // Avoid selecting the current level
-    if (rand_level_index >= rf::netgame.current_level_index) {
-        rand_level_index++;
+    // keep rolling until we get something that's not the current level
+    while (rf::netgame.levels[rand_level_index] == current_level_filename) {
+        rand_level_index = dist_levels(rng);
     }
```

Comment:
I do not like these changes. 1. It is simple code. `xlog` is not essential. 2. It loops now which is worse. 3. `selected_level_filename` but `filename` is fine. 4. `rf::netgame.level` is rf::String not `const char *`.

Reply:
> `xlog` is not essential

I was using this logging to identify why my random selection logic wasn't working properly before (it was due to me misunderstanding when `current_level_index` got updated), and I didn't remove the logging when I was done. Removed it now to clean up.

> It loops now which is worse.

If you have a better approach please do feel free to share, however from my read, this is the most elegant and efficient. I'll note that while it is a loop, it is very rare for it to even run twice, especially with a large server rotation.

> `selected_level_filename` but `filename` is fine.

Updated for simplicity.

> `rf::netgame.level` is rf::String not `const char *`.

This is totally my bad, thanks for noting it. Fixed, now the code returns a const char*, which is what `rf::multi_change_level` takes as input, but uses `rf::String` until then.

> Why not compute a `current_filename_index` instead?

This would result in `map_rand` diverging from how `level X` works - ie. it would make `map_rand` change the server's position in the rotation rather than simply swapping to a new map and then continuing with the rotation when the random map is done. That would be a net negative in my view, especially when a more intuitive way (the current approach) exists.